### DISCUSSION TITLE
Idea suggestion about timeout

### DISCUSSION THREAD

**Discussion by zion830** [euphony-io/euphony#186] [Ideas]
Hi! I have a new feature idea. 
Now in Euphony, we have to wait infinitely until communication finishes.
I think it will be useful to add a new function that sets the maximum waiting time, like the timeout option of HTTP.

### Interface
```Kotlin
// Set timeout
fun setTimeout(millisecond: Long, onTimeout: () -> Unit = {})

// Remove timeout setting
fun clearTimeout()
```
### Usage
```Kotlin
rxManager.setTimeout(1000L) {
    // Handle timeout event
}
rxManager.clearTimeout()
```
### Expected Exception
* When the millisecond value of `setTimeout()` is less than 0
    * In this case, I consider handling it in the same way that clearTimeout is called

**Comment by designe** [comment]
Hi! @zion830 Thanks for your interesting idea suggestion!
I like this feature! it will be very helpful to understand fail or success in seconds.

What I'm concern is that `interface` is really similar to `setTimeout` & `clearTimeout()` of javascript.
As you already know, those functions are timer API and web standard. 
So, Some people can misunderstand to use API. 

Let me give you some architecture decision candidates!

1. python `requests` https://linuxpip.org/python-requests-timeout/
2. jquery `ajax` https://api.jquery.com/jQuery.ajax/
3. axios https://github.com/axios/axios#note-commonjs-usage

From those cadidates, we can refer to implement timeout feature :)
How do you think about it?

**Reply by zion830** [reply]
Thank you for your feedback.
As you said, I referenced javascript's api to name methods. And I agree with your opinion.
I read documents you recommended, and I got another idea, which just add a new argument to `listen()`

For exmaple, like this way. 
```
fun listen(hex: Boolen = false, timeout: Long = 0L)
```
I think it is more simple :)

**Reply by designe** [reply]
Great conclusion :)
